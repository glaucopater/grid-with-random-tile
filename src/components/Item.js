import React, { Component, Fragment } from "react";
import { StyledItem, Content, Top, Right, Bottom, Left } from "./Styled";
import { Link } from "styled-icons/evil";

const StyledLink = () => {
  return (
    <Fragment>
      <Link size="16" />
    </Fragment>
  );
};

class Item extends Component {
  constructor(props) {
    super(props);
    this.state = {
      top: null,
      bottom: null,
      left: null,
      right: null,
      pos: { x: null, y: null },
      rightConnected: null,
      leftConnected: null,
      topConnected: null,
      bottomConnected: null
    };
  }

  render() {
    return (
      <StyledItem>
        <Content>
          <Top {...this.props}>{this.props.topConnected && <StyledLink />}</Top>
          <Right {...this.props}>
            {this.props.rightConnected && <StyledLink />}
          </Right>
          <Bottom {...this.props}>
            {this.props.bottomConnected && <StyledLink />}
          </Bottom>
          <Left {...this.props}>
            {this.props.leftConnected && <StyledLink />}
          </Left>
        </Content>
      </StyledItem>
    );
  }
}

export default Item;
