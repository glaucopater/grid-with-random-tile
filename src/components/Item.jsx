import React from "react";
import { 
  StyledItem, 
  Content, 
  Top, 
  Right, 
  Bottom, 
  Left,
  TopConnection,
  RightConnection,
  BottomConnection,
  LeftConnection
} from "./Styled";
import styled from "styled-components";

const IconWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 36px;
  height: 36px;
  border-radius: 50%;
  background-color: white;
  border: 2px solid #333;
  position: relative;
  z-index: 2;
  font-family: 'Roboto', sans-serif;
`;

const IconContent = styled.div`
  font-size: 24px;
  line-height: 1;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledLink = ({ style = {}, children }) => (
  <IconWrapper style={{ ...style }}>
    <IconContent>
      {children || "ðŸ”—"}
    </IconContent>
  </IconWrapper>
);

const StyledBlock = ({ style = {}, children }) => (
  <IconWrapper style={{ ...style }}>
    <IconContent>
      {children || "â›”"}
    </IconContent>
  </IconWrapper>
);

const DirectionalComponent = ({ isConnected, linkProps = {}, blockProps = {} }) =>
  isConnected ? <StyledLink {...linkProps} /> : <StyledBlock {...blockProps} />;

const getRandomColor = () => {
  const letters = '0123456789ABCDEF';
  let color = '#';
  for (let i = 0; i < 6; i++) {
    color += letters[Math.floor(Math.random() * 16)];
  }
  return color;
};

const TileNumber = styled.div`
  position: absolute;
  top: 5px;
  left: 5px;
  font-size: 14px;
  font-weight: bold;
  color: #333;
  background-color: rgba(255, 255, 255, 0.7);
  border-radius: 50%;
  width: 20px;
  height: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: 'Roboto', sans-serif;
`;

const Item = ({ firstAction, topConnected, rightConnected, bottomConnected, leftConnected, pos, index, children, ...props }) => {
  const color = React.useMemo(() => getRandomColor(), []);
  const hasConnections = topConnected || rightConnected || bottomConnected || leftConnected;

  return (
    <StyledItem color={color} style={{ opacity: hasConnections ? 1 : 0.5, fontFamily: 'Roboto, sans-serif' }}>
      <TileNumber>{children}</TileNumber>
      {topConnected && <TopConnection />}
      {rightConnected && <RightConnection />}
      {bottomConnected && <BottomConnection />}
      {leftConnected && <LeftConnection />}
      <Content>
        <Top>
          <DirectionalComponent isConnected={topConnected} />
        </Top>
        <Right>
          <DirectionalComponent isConnected={rightConnected} />
        </Right>
        <Bottom>
          <DirectionalComponent isConnected={bottomConnected} />
        </Bottom>
        <Left>
          <DirectionalComponent isConnected={leftConnected} />
        </Left>
      </Content>
    </StyledItem>
  );
};

export default Item;
