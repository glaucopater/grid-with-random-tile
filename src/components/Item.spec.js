import React, { Component } from "react";
import Item from "./Item";
import { shallow } from "enzyme";

import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({ adapter: new Adapter() });

it("renders properly", () => {
  const wrapper = shallow(<Hello name="CodeSandbox" />);
  expect(wrapper.text()).toEqual("Hello CodeSandbox!");
});

describe("convertBlock", () => {
  it("should render an item", () => {
    const renderedComponent = shallow(<Item />);
    expect(renderedComponent).toBe(true);
  });
});
