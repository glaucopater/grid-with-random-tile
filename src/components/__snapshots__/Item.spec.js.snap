// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<styled.div>
  <styled.div>
    <Styled(styled.div) />
    <Styled(styled.div) />
    <Styled(styled.div) />
    <Styled(styled.div) />
  </styled.div>
</styled.div>
`;
